File: web/src/views/traffic.ts
// What the capture is made of. Every row of the waterfall is one 802.11 frame that happened to
// be on the air, so the sample rate of every analysis in this app is a property of the room's
// traffic rather than of the server — and when breathing will not converge or the waterfall grows
// stripes, this is the page that says why.
//
// Two numbers do most of the work here, and they are not the same number. The rate is how many
// frames a second arrived; the duty cycle is how many of those seconds carried any at all. A link
// at 40 Hz and duty 1.0 draws an even waterfall and resamples cleanly. A link averaging 40 Hz in
// bursts, silent half the time, produces the identical rate and an unusable window — `Coverage`
// in the DSP layer exists precisely because interpolating across those holes puts invented power
// in the respiration band.
//
// The breakdown by transmitter is what makes it actionable. A node filtered to one BSSID lives
// and dies by that access point's beacon rate; an unfiltered monitor is sampling whatever the
// neighbours are doing, and seeing which MAC actually feeds the waterfall is how you decide
// whether to change channel, lock to another AP, or turn the stimulus on.

import { clear, el, fitCanvas, formatAgo, hintBlock, stat, viewLayout } from "../lib/dom";
import type { TrafficNode, TrafficSource } from "../lib/messages";
import { SERIES_COLORS, beginPlot, drawFrame, plotArea } from "../lib/plot";
import { store } from "../lib/store";
import type { View } from "./view";

// Seconds between polls. The server buckets by the second, so anything faster would redraw the
// same numbers; anything slower would make a stall that has just started look like it is still
// going when it has already cleared.
const POLL_MS = 1000;

// How many transmitters are listed before the rest are folded away. A busy channel produces a
// long tail of one-frame passers-by, and the whole page is about which sources matter.
const VISIBLE_SOURCES = 12;

// Sources beyond the palette, and the frames no source claimed, are drawn in this. Deliberately
// the muted grey rather than another hue: "not individually interesting" is the message.
const REST_COLOR = "#5b6474";

// The per-row sparkline, in CSS pixels. Sized here rather than in the stylesheet because the
// canvas is drawn the moment it is built, before it has been laid out — measuring its box would
// mean deferring to an animation frame, and a hidden tab runs no animation frames at all, so
// every row would stay blank until the tab came back.
const SPARK_W = 84;
const SPARK_H = 26;

function verdict(node: TrafficNode): { text: string; className: string } {
  if (node.frames_window === 0) return { text: "silent", className: "bad" };
  // Ranked above the rate checks on purpose. An intermittent capture is not a slow capture: the
  // fix is different (find out what stops the traffic) and the failure is worse, because the mean
  // rate looks fine while every analysis window has a hole in it.
  if (node.duty < 0.5) return { text: "intermittent", className: "bad" };
  if (node.max_gap_s > 1) return { text: "stalling", className: "warn" };
  if (node.rate_hz < 20) return { text: "starved", className: "warn" };
  if (node.rate_hz < 40) return { text: "thin", className: "warn" };
  return { text: "steady", className: "good" };
}

function label(source: TrafficSource): string {
  if (source.ap) return source.ssid ? `AP · ${source.ssid}` : "AP";
  return "station";
}

function rssiText(source: TrafficSource): string {
  if (source.rssi_mean === null) return "—";
  const spread =
    source.rssi_min !== null && source.rssi_max !== null && source.rssi_min !== source.rssi_max
      ? ` (${source.rssi_min}…${source.rssi_max})`
      : "";
  return `${source.rssi_mean.toFixed(0)} dBm${spread}`;
}

export function trafficView(): View {
  const summary = el("div", { class: "stat-row" });
  const verdictPill = el("span", { class: "pill" }, "—");
  const canvas = el("canvas", { class: "plot-canvas traffic-canvas" });
  const caption = el("p", { class: "panel-note" });
  const sourceList = el("div", { class: "traffic-sources" });
  const sourceHead = el("div", { class: "wifi-subhead" }, el("h3", {}, "Transmitters"));
  const notes = el("div", { class: "traffic-notes" });

  let current: TrafficNode | null = null;
  let pollTimer: number | null = null;
  let showAll = false;

  // Colour per MAC, assigned in the order sources are first seen and kept for as long as the view
  // is on one node. Colouring by rank instead would swap two sources' colours whenever their
  // frame counts crossed, which on a busy channel is every few seconds.
  const colors = new Map<string, string>();
  let coloredNode: number | null = null;

  function colorFor(mac: string): string {
    const existing = colors.get(mac);
    if (existing !== undefined) return existing;
    const next = colors.size < SERIES_COLORS.length ? SERIES_COLORS[colors.size] : REST_COLOR;
    colors.set(mac, next);
    return next;
  }

  function render(nodes: TrafficNode[]) {
    const nodeId = store.selectedNode.value;
    const node = nodes.find((candidate) => candidate.node_id === nodeId) ?? null;

    if (node !== null && node.node_id !== coloredNode) {
      // A different node hears a different set of transmitters; carrying the assignment across
      // would hand the new node's loudest source whatever colour the old one's had.
      colors.clear();
      coloredNode = node.node_id;
      showAll = false;
    }
    current = node;

    renderSummary(node);
    renderSources(node);
    renderNotes(node);
    draw();
  }

  function renderSummary(node: TrafficNode | null) {
    clear(summary);
    if (node === null) {
      verdictPill.className = "pill";
      verdictPill.textContent = "no frames";
      summary.append(
        el(
          "p",
          { class: "hint" },
          "Nothing has arrived from this node. The traffic breakdown is built from the frames " +
            "themselves, so it appears as soon as the first one does.",
        ),
      );
      return;
    }

    const state = verdict(node);
    verdictPill.className = `pill pill-${state.className}`;
    verdictPill.textContent = state.text;

    const stations = node.sources.filter((source) => !source.ap).length;
    const aps = node.sources.length - stations;

    summary.append(
      stat("capture rate", `${node.rate_hz.toFixed(1)} Hz`, node.rate_hz < 40 ? "warn" : ""),
      // Rounded down rather than to nearest: 99.6% displayed as 100% on a link that dropped a
      // second is the one rounding error that matters here.
      stat("duty", `${Math.floor(node.duty * 100)}%`, node.duty < 0.95 ? "warn" : ""),
      stat(
        "longest gap",
        node.max_gap_s >= 0.05 ? `${node.max_gap_s.toFixed(2)} s` : "—",
        node.max_gap_s > 1 ? "bad" : node.max_gap_s > 0.25 ? "warn" : "",
      ),
      stat("transmitters", `${node.sources.length}`),
      stat("access points", `${aps}`),
      stat("frames", node.frames_window.toLocaleString()),
    );

    caption.textContent =
      `frames per second over the last ${node.window_s} s, stacked by transmitter · ` +
      `${node.silent_s} s of the ${node.covered_s} s measured carried nothing`;
  }

  function renderSources(node: TrafficNode | null) {
    clear(sourceList);
    if (node === null) return;

    if (node.sources.length === 0) {
      sourceList.append(
        el(
          "p",
          { class: "hint" },
          node.anonymous > 0
            ? "Every frame from this node arrived without a source address, which is what a v1 " +
              "uplink looks like — the field was added in v2. The rate and duty above are still " +
              "measured; only the attribution is missing."
            : "No transmitters yet.",
        ),
      );
      return;
    }

    const peak = Math.max(1, ...node.counts);
    const shown = showAll ? node.sources : node.sources.slice(0, VISIBLE_SOURCES);
    for (const source of shown) sourceList.append(renderSource(source, peak));

    if (node.sources.length > shown.length) {
      sourceList.append(
        el(
          "button",
          {
            class: "button button-wide",
            onclick: () => {
              showAll = true;
              renderSources(current);
            },
          },
          `Show all ${node.sources.length}`,
        ),
      );
    }
  }

  function renderSource(source: TrafficSource, peak: number): HTMLElement {
    const color = colorFor(source.mac);
    const ratio = window.devicePixelRatio || 1;
    const spark = el("canvas", {
      class: "traffic-spark",
      width: Math.round(SPARK_W * ratio),
      height: Math.round(SPARK_H * ratio),
      style: `width:${SPARK_W}px;height:${SPARK_H}px`,
    }) as HTMLCanvasElement;
    drawSparkline(spark, source.counts, peak, color);

    return el(
      "div",
      { class: "traffic-source" },
      el("span", { class: "traffic-swatch", style: `background:${color}` }),
      el(
        "div",
        { class: "traffic-source-body" },
        el(
          "div",
          { class: "traffic-source-head" },
          el("span", { class: "traffic-mac" }, source.mac),
          el("span", { class: `pill ${source.ap ? "pill-good" : ""}` }, label(source)),
          source.randomized
            ? el(
                "span",
                {
                  class: "pill pill-warn",
                  title:
                    "Locally administered address: a randomized MAC. It will be a different " +
                    "address next time this device associates.",
                },
                "randomized",
              )
            : null,
          source.malformed
            ? el(
                "span",
                {
                  class: "pill pill-bad",
                  title:
                    "A group bit in a source address, which cannot happen on the air. The " +
                    "node is reading the MAC from the wrong offset.",
                },
                "malformed",
              )
            : null,
        ),
        el(
          "div",
          { class: "traffic-stats" },
          el("span", { class: "traffic-rate" }, `${source.rate_hz.toFixed(1)} Hz`),
          el("span", {}, `${Math.round(source.share * 100)}% of frames`),
          el("span", {}, rssiText(source)),
          el("span", { class: "traffic-link" }, `${source.n_sub} sub · ch ${source.channel || "—"}`),
          el(
            "span",
            { class: source.max_gap_s > 1 ? "warn" : "" },
            source.max_gap_s >= 0.05 ? `gap ${source.max_gap_s.toFixed(2)} s` : "no gaps",
          ),
          el("span", { class: "traffic-ago" }, formatAgo(source.last_seen)),
        ),
        el(
          "div",
          { class: "traffic-bar" },
          el("span", {
            style: `width:${Math.max(1, source.share * 100)}%;background:${color}`,
          }),
        ),
      ),
      spark,
    );
  }

  function renderNotes(node: TrafficNode | null) {
    clear(notes);
    if (node === null) return;

    const lines: (Node | string)[] = [];

    if (node.sources.length === 1 && node.anonymous === 0) {
      lines.push(
        "One transmitter only. Either the channel is quiet, or this node is filtered to a " +
          "single BSSID (CSI_AP_MAC), in which case the capture rate is that one access " +
          "point's frame rate and nothing else on the channel can lift it.",
      );
    }
    if (node.anonymous > 0 && node.sources.length > 0) {
      lines.push(
        `${node.anonymous.toLocaleString()} frames carried no source address (a v1 uplink) and ` +
          "are counted in the totals but attributed to nobody.",
      );
    }
    if (node.evicted > 0) {
      lines.push(
        `${node.evicted.toLocaleString()} quiet transmitters have been forgotten to keep the ` +
          "list bounded — an unfiltered monitor hears a long tail of one-off addresses.",
      );
    }
    if (node.rate_hz < 40 && node.frames_window > 0) {
      lines.push(
        el(
          "span",
          {},
          "Below about 40 Hz the heart-rate band is the first thing to go. Turning on traffic " +
            "generation in ",
          el("a", { href: "#wifi" }, "WiFi"),
          " makes the node provoke frames of its own instead of waiting for the household's.",
        ),
      );
    }

    for (const line of lines) notes.append(el("p", { class: "hint" }, line));
  }

  /**
   * The stacked timeline: one column per second, each split by transmitter.
   *
   * Stacked rather than one line per source because the quantity being explained is the total —
   * how many rows a second the waterfall got — and the interesting question is which transmitter
   * the total is made of, and what disappears when it dips.
   */
  function draw() {
    fitCanvas(canvas);
    const plot = beginPlot(canvas);
    if (plot === null) return;

    const node = current;
    if (node === null || node.counts.length === 0) {
      drawFrame(plot, [], []);
      return;
    }

    const area = plotArea(plot);
    const { context } = plot;
    const n = node.counts.length;
    const peak = Math.max(1, ...node.counts);
    const columnWidth = area.w / n;

    // The stretch before this node said anything is not a stall — it is a period we were not
    // listening in — so it is shaded rather than left as an indistinguishable run of zeroes.
    const listeningFrom = n - node.covered_s;
    if (listeningFrom > 0) {
      context.fillStyle = "rgba(255,255,255,0.03)";
      context.fillRect(area.x0, area.y0, listeningFrom * columnWidth, area.h);
    }

    const stacked = new Float64Array(n);
    for (const source of node.sources) {
      context.fillStyle = colorFor(source.mac);
      for (let i = 0; i < n; i++) {
        const count = source.counts[i];
        if (!count) continue;
        const height = (count / peak) * area.h;
        const y = area.y1 - ((stacked[i] / peak) * area.h + height);
        context.fillRect(area.x0 + i * columnWidth, y, Math.max(1, columnWidth - 0.5), height);
        stacked[i] += count;
      }
    }

    // Whatever the sources did not account for: frames with no source address. Drawn last, on
    // top, so a v1 node still gets a timeline instead of an empty plot.
    context.fillStyle = REST_COLOR;
    for (let i = 0; i < n; i++) {
      const remainder = node.counts[i] - stacked[i];
      if (remainder <= 0) continue;
      const height = (remainder / peak) * area.h;
      const y = area.y1 - ((stacked[i] / peak) * area.h + height);
      context.fillRect(area.x0 + i * columnWidth, y, Math.max(1, columnWidth - 0.5), height);
    }

    drawFrame(
      plot,
      [
        [0, `-${node.window_s}s`],
        [0.5, `-${Math.round(node.window_s / 2)}s`],
        [1, "now"],
      ],
      [
        [0, "0"],
        [0.5, String(Math.round(peak / 2))],
        [1, `${peak} Hz`],
      ],
    );
  }

  function drawSparkline(
    spark: HTMLCanvasElement,
    counts: number[],
    peak: number,
    color: string,
  ) {
    const context = spark.getContext("2d");
    if (context === null) return;
    // Normalized to the node's busiest second, not to this source's own, so the rows are
    // comparable: a quiet transmitter should look quiet next to the one carrying the capture.
    const columnWidth = spark.width / counts.length;
    context.fillStyle = color;
    for (let i = 0; i < counts.length; i++) {
      if (!counts[i]) continue;
      // Two device pixels minimum. A second that carried a couple of frames is a whole column
      // here, and rounded down to a sub-pixel sliver it disappears entirely — which would make
      // a source that is quietly still there look like one that stopped.
      const height = Math.max(2, (counts[i] / peak) * spark.height);
      context.fillRect(i * columnWidth, spark.height - height, Math.max(1, columnWidth), height);
    }
  }

  const root = viewLayout({
    className: "view-traffic",
    main: [
      el(
        "div",
        { class: "panel" },
        el(
          "div",
          { class: "panel-head" },
          el("h2", {}, "Capture"),
          verdictPill,
        ),
        summary,
        el("div", { class: "plot-frame traffic-plot" }, canvas),
        caption,
        hintBlock(
          "Every row of the waterfall is one frame that was on the air, so this is the sample " +
            "rate every other view inherits. Rate and duty are different faults: 40 Hz evenly " +
            "spread resamples cleanly, while 40 Hz in bursts leaves holes that the DSP layer " +
            "refuses to interpolate across — which is why breathing can go quiet on a link " +
            "whose average rate looks perfectly healthy.",
        ),
      ),
      el("div", { class: "panel panel-grow" }, sourceHead, sourceList, notes),
    ],
  });

  const unsubscribes: (() => void)[] = [];
  const onResize = () => draw();

  return {
    id: "traffic",
    title: "Traffic",
    root,
    mount() {
      unsubscribes.push(store.traffic.subscribe(render));
      unsubscribes.push(store.selectedNode.subscribe(() => render(store.traffic.value), false));
      window.addEventListener("resize", onResize);
      void store.refreshTraffic();
      pollTimer = window.setInterval(() => void store.refreshTraffic(), POLL_MS);
    },
    unmount() {
      while (unsubscribes.length > 0) unsubscribes.pop()?.();
      window.removeEventListener("resize", onResize);
      if (pollTimer !== null) {
        window.clearInterval(pollTimer);
        pollTimer = null;
      }
    },
  };
}

File: server/csi/traffic.py
"""What the capture is actually made of: which transmitters, at what rate, with what holes.

Every row of the waterfall is one 802.11 frame that happened to be on the air. That is the whole
of the sampling story — the analysis rate is not something the server picks, it is whatever the
room's traffic handed over. So when breathing will not converge or the waterfall grows stripes,
the question is almost never "is the DSP wrong" and almost always "what was the air doing", and
until now the only answer available was a cumulative frame count per MAC with no time axis.

This module keeps the time axis. Per node and per source MAC it holds one-second buckets in a
fixed ring, which is what makes "80 Hz" separable from "800 frames in one second and silence
for the next nine" — two captures with the same mean rate, one of which is usable and one of
which is not.

Buckets rather than a deque of arrival times: at 100 Hz a two-minute window is 12000 timestamps
per source, and the report walks every source at 1 Hz. The bucket ring is 120 slots whatever the
rate, it never allocates after construction, and every number the UI shows — rate, duty cycle,
longest gap, RSSI spread, the sparkline itself — is a walk over those slots.

No I/O and no asyncio here: `Hub.handle_datagram` calls `observe` on the event loop (a handful
of integer operations per frame) and the HTTP handler calls `report`. Nothing in this module
touches a ring, so invariant 4 is not in play.
"""

from __future__ import annotations

from collections.abc import Mapping

from .protocol import Frame

# Seconds of per-second history kept, per node and per source. Matches the default in-memory CSI
# history, so the traffic timeline covers the same span as the waterfall it explains.
WINDOW_S = 120

# Distinct source MACs remembered per node. An unfiltered monitor on a busy channel hears a long
# tail of one-off addresses — probe requests, passers-by, a neighbour's phone waking up — and
# past this many, the longest-silent are the right ones to forget.
MAX_SOURCES = 64


def format_mac(mac: bytes) -> str:
    return mac.hex(":")


def is_locally_administered(mac: bytes) -> bool:
    """Bit 1 of the first octet: the address was assigned by software, not by a vendor.

    On a modern phone or laptop this is what MAC randomization looks like from the air. Worth
    surfacing because it explains a source that appears, contributes a burst of frames and never
    comes back under the same name — which otherwise reads as a flapping link.
    """
    return bool(mac[0] & 0x02)


def is_group(mac: bytes) -> bool:
    """Bit 0 of the first octet: a multicast or broadcast address.

    Never legal in a *source* address, so seeing one here does not mean interesting traffic — it
    means the offset this MAC was read from is wrong. Reported rather than hidden, because the
    symptom of a misread header is otherwise a plausible-looking table of transmitters that do
    not exist (see the header-size argument in pi/csi_node.py).
    """
    return bool(mac[0] & 0x01)


class SecondBuckets:
    """One-second counters in a fixed ring, addressed by absolute second.

    Each slot remembers which second it holds. A slot whose second falls outside the window is
    stale by definition and reads as zero, so nothing has to expire anything on a timer — an
    idle source simply stops being counted, and one that goes quiet for an hour costs the same
    as one that never stopped.
    """

    __slots__ = ("size", "_second", "_count", "_rssi_sum", "_rssi_min", "_rssi_max", "_gap_max")

    def __init__(self, size: int = WINDOW_S) -> None:
        self.size = size
        self._second = [-1] * size
        self._count = [0] * size
        self._rssi_sum = [0] * size
        self._rssi_min = [0] * size
        self._rssi_max = [0] * size
        self._gap_max = [0.0] * size

    def add(self, received_at: float, rssi: int, gap_s: float) -> None:
        second = int(received_at)
        slot = second % self.size
        if self._second[slot] != second:
            self._second[slot] = second
            self._count[slot] = 0
            self._rssi_sum[slot] = 0
            self._rssi_min[slot] = rssi
            self._rssi_max[slot] = rssi
            self._gap_max[slot] = 0.0
        self._count[slot] += 1
        self._rssi_sum[slot] += rssi
        if rssi < self._rssi_min[slot]:
            self._rssi_min[slot] = rssi
        if rssi > self._rssi_max[slot]:
            self._rssi_max[slot] = rssi
        if gap_s > self._gap_max[slot]:
            self._gap_max[slot] = gap_s

    def window(self, now: float) -> WindowSummary:
        """Everything derivable from the last `size` seconds, in one pass."""
        newest = int(now)
        oldest = newest - self.size + 1
        counts = [0] * self.size
        frames = 0
        rssi_sum = 0
        rssi_min: int | None = None
        rssi_max: int | None = None
        gap_max = 0.0
        active = 0

        for slot in range(self.size):
            second = self._second[slot]
            if second < oldest or second > newest:
                continue
            count = self._count[slot]
            if count == 0:
                continue
            counts[second - oldest] = count
            frames += count
            active += 1
            rssi_sum += self._rssi_sum[slot]
            low, high = self._rssi_min[slot], self._rssi_max[slot]
            rssi_min = low if rssi_min is None or low < rssi_min else rssi_min
            rssi_max = high if rssi_max is None or high > rssi_max else rssi_max
            if self._gap_max[slot] > gap_max:
                gap_max = self._gap_max[slot]

        return WindowSummary(counts, frames, rssi_sum, rssi_min, rssi_max, gap_max, active)


class WindowSummary:
    """One `SecondBuckets` read, kept as an object so callers do not unpack a seven-tuple."""

    __slots__ = ("counts", "frames", "rssi_sum", "rssi_min", "rssi_max", "gap_max", "active_s")

    def __init__(
        self,
        counts: list[int],
        frames: int,
        rssi_sum: int,
        rssi_min: int | None,
        rssi_max: int | None,
        gap_max: float,
        active_s: int,
    ) -> None:
        self.counts = counts
        self.frames = frames
        self.rssi_sum = rssi_sum
        self.rssi_min = rssi_min
        self.rssi_max = rssi_max
        self.gap_max = gap_max
        self.active_s = active_s

    @property
    def rssi_mean(self) -> float | None:
        return self.rssi_sum / self.frames if self.frames else None


class Source:
    """One transmitter as heard by one node."""

    __slots__ = (
        "mac",
        "frames",
        "first_seen",
        "last_seen",
        "rssi",
        "channel",
        "n_sub",
        "buckets",
        "_last_arrival",
    )

    def __init__(self, mac: bytes, received_at: float, window_s: int) -> None:
        self.mac = mac
        self.frames = 0
        self.first_seen = received_at
        self.last_seen = received_at
        self.rssi = 0
        self.channel = 0
        self.n_sub = 0
        self.buckets = SecondBuckets(window_s)
        self._last_arrival = 0.0

    def observe(self, frame: Frame, received_at: float) -> None:
        gap = received_at - self._last_arrival if self._last_arrival else 0.0
        self._last_arrival = received_at
        self.last_seen = received_at
        self.frames += 1
        self.rssi = frame.rssi
        self.channel = frame.channel
        self.n_sub = frame.n_sub
        self.buckets.add(received_at, frame.rssi, gap)


class TrafficTracker:
    """The traffic picture for one node.

    Not cleared on a roam or a reboot, unlike everything else the hub keeps per node. Those
    events clear history because a baseline measured on the old link describes a different room;
    this measures the air rather than the room, and a re-association is precisely the thing you
    open this view to see — the old BSSID going quiet at the same second the new one starts.
    """

    def __init__(self, window_s: int = WINDOW_S, max_sources: int = MAX_SOURCES) -> None:
        self.window_s = window_s
        self.max_sources = max_sources
        self.sources: dict[bytes, Source] = {}
        self.buckets = SecondBuckets(window_s)
        self.frames = 0
        # Frames whose source MAC is six zero bytes: a v1 uplink, which has no such field. They
        # count towards the node's rate — they are real captured frames — but they cannot be
        # attributed, and pretending 00:00:00:00:00:00 is a transmitter would put a phantom row
        # at the top of the table for every ESP32 running older firmware.
        self.anonymous = 0
        self.evicted = 0
        self.first_seen = 0.0
        self.last_seen = 0.0
        self._last_arrival = 0.0

    def observe(self, frame: Frame, received_at: float) -> None:
        if not self.first_seen:
            self.first_seen = received_at
        gap = received_at - self._last_arrival if self._last_arrival else 0.0
        self._last_arrival = received_at
        self.last_seen = received_at
        self.frames += 1
        self.buckets.add(received_at, frame.rssi, gap)

        mac = frame.src_mac
        if not any(mac):
            self.anonymous += 1
            return

        source = self.sources.get(mac)
        if source is None:
            if len(self.sources) >= self.max_sources:
                quietest = min(self.sources, key=lambda key: self.sources[key].last_seen)
                del self.sources[quietest]
                self.evicted += 1
            source = self.sources[mac] = Source(mac, received_at, self.window_s)
        source.observe(frame, received_at)

    # -- reporting -------------------------------------------------------------------------

    def transmitters(self) -> list[dict]:
        """The WiFi overview's shape: cumulative totals, loudest first.

        Kept alongside `report` rather than derived from it because that page wants the long
        view — everything this node has ever heard — while the traffic view wants the window.
        """
        return [
            {
                "mac": format_mac(source.mac),
                "frames": source.frames,
                "rssi": source.rssi,
                "channel": source.channel,
                "last_seen": source.last_seen,
            }
            for source in sorted(self.sources.values(), key=lambda s: -s.frames)
        ]

    def report(self, now: float, aps: Mapping[str, str] | None = None) -> dict:
        """The full picture for one node over the last `window_s` seconds.

        `aps` maps BSSID to SSID, from the node's last scan. It is what lets a row be labelled
        as an access point rather than left as an anonymous MAC, and it is passed in rather than
        looked up here so this module stays free of the control store.
        """
        node = self.buckets.window(now)
        covered = self._covered_seconds(now)
        span_s = float(max(covered, 1))

        sources = [
            self._source_report(source, now, node.frames, aps or {})
            for source in self.sources.values()
        ]
        # Loudest in the window first, then by recency — so a source that has gone silent sinks
        # rather than holding a top row on the strength of what it did two minutes ago.
        sources.sort(key=lambda row: (-row["frames_window"], -row["last_seen"]))

        return {
            "node_id": None,  # filled in by the caller, which is the one that knows
            "window_s": self.window_s,
            "covered_s": covered,
            "counts": node.counts,
            "frames": self.frames,
            "frames_window": node.frames,
            "rate_hz": node.frames / span_s,
            # Fraction of the covered seconds that carried at least one frame. This is the
            # number that separates a slow capture from an intermittent one: 40 Hz at duty 1.0
            # draws an even waterfall, 40 Hz at duty 0.5 draws it in stripes.
            "duty": node.active_s / covered if covered else 0.0,
            "silent_s": max(0, covered - node.active_s),
            "max_gap_s": node.gap_max,
            "anonymous": self.anonymous,
            "evicted": self.evicted,
            "first_seen": self.first_seen,
            "last_seen": self.last_seen,
            "sources": sources,
        }

    def _covered_seconds(self, now: float) -> int:
        """How much of the window this node has actually been around for.

        A node that appeared three seconds ago has not been silent for the other 117, and
        dividing by the full window would report it at a fortieth of its real rate.
        """
        if not self.first_seen:
            return 0
        seen = int(now) - int(self.first_seen) + 1
        return max(1, min(self.window_s, seen))

    def _source_report(
        self, source: Source, now: float, node_frames: int, aps: Mapping[str, str]
    ) -> dict:
        window = source.buckets.window(now)
        covered = max(1, min(self.window_s, int(now) - int(source.first_seen) + 1))
        mac = format_mac(source.mac)
        ssid = aps.get(mac)
        return {
            "mac": mac,
            "frames": source.frames,
            "frames_window": window.frames,
            "rate_hz": window.frames / covered,
            "share": window.frames / node_frames if node_frames else 0.0,
            "counts": window.counts,
            "rssi": source.rssi,
            "rssi_mean": window.rssi_mean,
            "rssi_min": window.rssi_min,
            "rssi_max": window.rssi_max,
            "channel": source.channel,
            "n_sub": source.n_sub,
            "max_gap_s": window.gap_max,
            "duty": window.active_s / covered,
            "first_seen": source.first_seen,
            "last_seen": source.last_seen,
            # An access point if the node's own scan saw this BSSID. Anything else is a station
            # — or rather, is not known to be an access point, which is a weaker claim and the
            # only honest one available from a source address alone.
            "ap": ssid is not None,
            "ssid": ssid or "",
            "randomized": is_locally_administered(source.mac),
            "malformed": is_group(source.mac),
        }
